README.md
# Belle

Configurable React Components with great UX.

Website & Documentation: [http://nikgraf.github.io/belle/](http://nikgraf.github.io/belle/)

[<img src="https://cloud.githubusercontent.com/assets/223045/8707250/f77634f8-2b34-11e5-92d1-f0b0fe44ade9.png" width="50%">](http://nikgraf.github.io/belle/)

[![Build Status](https://travis-ci.org/nikgraf/belle.svg)](https://travis-ci.org/nikgraf/belle)
[![Dependency Status](https://david-dm.org/nikgraf/belle.svg)](https://david-dm.org/nikgraf/belle)
[![peerDependency Status](https://david-dm.org/nikgraf/belle/peer-status.svg)](https://david-dm.org/nikgraf/belle#info=peerDependencies)

## Getting Started

Belle is available as [npm](http://npmjs.org) package. Once you have npm you can install Belle in your project folder with:

```
npm install belle
```

### Import & use Belle Components

We recommend you to get started with [React](https://facebook.github.io/react/) first. Once you have a simple application setup you can import any Belle component and use it right away. No stylesheets, font or any other prerequisite needed.

```html
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
</head>
<body>
  <div id="react-root"></div>
  <!--
    You can use browserify, webpack or similar tools
    to compile & combine your JSX code
  -->
  <script src="bundle.js"></script>
</body>
</html>
```

```javascript
var React = require('react');
var belle = require('belle');
var TextInput = belle.TextInput;

var App = React.createClass({

  render: function() {
    return (
      <div>
        <TextInput defaultValue="Update here and see how the input grows …" />
      </div>
    );
  }
});

React.render(<App/>, document.getElementById('react-root'));
```

### Learn more

In addition you can dig through the [documentation](http://nikgraf.github.io/belle/) to learn about how to modify Belle components.

## Browser Support

- Chrome (mobile and desktop)
- Safari (mobile and desktop)
- Firefox
- Internet Explorer 9, 10, 11

## Philosophy

"Great UX starts with a good UI"

With Belle I'm aiming to provide a library that follows a couple well established
principles. While some of them might not be consider best practice today
We believe they will become some day in the near future.

### Easy to use

The most important attribute of any great experience is that there was no hassle to achieve
the initial goals.

### Consistent Behavior

Every object someone interacts with has it's own little language. This language
must be learned by everyone getting in touch with a new set of object. In order
to provide a great experience the amount to learn should be reduced to a minimum.

There are two major personalities to target with Belle. One is the developer.
For him/her APIs should be provided in a consistent way through all components.
The other and more important is the users of the components. Every color,
animation or behavior should be aligned with the other components to provide
a great experience.

### Encapsulate Styles

There is no reason while a style designed one specific element should affect
others. In CSS styles are often defined by overwriting previous ones and introducing
deeper and deeper nesting. Once nesting is introduced resolution of styles for one
specific element is not a trivial task anymore. Due this managing CSS dependencies
is hard. It is hard to predict how an application looks like after updating or
removing dependency.

That's why with Belle styles should apply to the components themselves in the DOM.
By doing so the visual appearance and business logic are combined in location.

### Every Interaction is followed up with Feedback

Let people know how their behavior affects the system. It assures them that
their input was acknowledged which provides them with a feeling of control over
the system.

### High Performance

The user should see affects of their actions instantly. Any delay can cause confusion
and frustration. While instant certainly is not always possible Belle strives to
provide an experience close to instant.

## Development

You can install the development environment with

```
npm install
```

`npm run build` will trigger a build into the `lib` folder. To develop a component it's convenient to use the examples or docs application.

### Run the examples or docs application

To run the examples or docs you go into the folder `docs` or `examples` and run `npm install` and `npm start`. The app will run with hot reloading on `http://localhost:3000`.

### Tests

In order to run the tests use

```
npm test
```

To run the test continuously you can use `npm run test:watch`.

## Discussion or need help?

In addition you can ask the community by [posting your question to StackOverflow with the **belle** tag.](http://stackoverflow.com/questions/ask?tags=belle).

## Future Plans

- Components to add: Dateformatter, Datepicker, Tooltip, Popover, Modal, Navigation Menu, NumberInput, EmailInput, Anchor, DropZone

## Special Thanks

Thanks to [Andrey Popp](https://github.com/andreypopp) & [Eugene](https://github.com/eugene1g) for their inspiring work on [React TextArea Autosize](https://github.com/andreypopp/react-textarea-autosize) which kind of ignited the idea of Belle.

Thanks to Christian Steiner from [http://www.cropd.at/](http://cropd.at) for creating the logo and helping out with the design.

Special thanks to [Jyoti Puri](https://github.com/jpuri) for the tremendous amount of work she put into this endeavor.

## License

MIT


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## To Be Released

## 4.0.0 - 2017-10-10

- Support React 15.4.0 [#341](https://github.com/nikgraf/belle/pull/341)

## 3.0.2 - 2017-05-30

- Removed duplicate declaration of style property in component Day [#336](https://github.com/nikgraf/belle/pull/336)

## 3.0.1 - 2017-01-16

- [Select] Fixed issue with Internet Explorer 11. The onBlur event is triggered before the onClick event occurs. Using onMouseDown resolved it. [#333](https://github.com/nikgraf/belle/pull/333)

## 3.0.0 - 2016-12-05

- [All Components] Fixed incompatible with React v15.4.0 [#330](https://github.com/nikgraf/belle/issues/330)

## 2.0.9 - 2016-10-21

- [DatePicker, ActionArea] Avoid firing the ActionArea's onUpdate twice on touch devices e.g. next month button & previous month button.

## 2.0.8 - 2016-10-17

- [All Components] Fixed various warnings introduced throughout React 15.x

## 2.0.7 - 2016-04-05

Due connection issues to NPM I republished it.

## 2.0.6 - 2016-04-05

### Fixed
- [Rating] Fixed an issue in Safari where selecting a rating was not possible [#277](https://github.com/nikgraf/belle/issues/277)

## 2.0.5 - 2016-03-28

### Changed
- [All Components] Changed internal ID generation to make Belle compatible to React 15.

### Fixed
- [Combobox] Clicking on an filtered option now picks the correct entry [#269](https://github.com/nikgraf/belle/issues/269)

## 2.0.4 - 2016-02-16

### Changed
- [Button, ComboBox, DatePicker, Rating, Select, TextInput, Toggle] Improved rendering performance by binding the functions to the class instantiation instead of every render due binding in JSX. Find more info [here](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md#lists-of-items) [#263](https://github.com/nikgraf/belle/pull/263) & [#265](https://github.com/nikgraf/belle/pull/265)

### Fixed

- [DatePicker] use nextMonthNavProps.className for next month (was prevMonthNavProps before) [b9937bb](https://github.com/nikgraf/belle/commit/b9937bbb96925899cb48bf55558119ad72cf2021)
- [DatePicker] fix broken Keyboard navigation in Firefox 39.0 and properly parse Date when jumping from January 2016 to December 2015 via keyboard in Chrome 48 [#264](https://github.com/nikgraf/belle/pull/264)

## 2.0.3 - 2015-12-07

### Fixed

- [Select] Ensure the exact same markup is generated every time the element is rendered [#155](https://github.com/nikgraf/belle/issues/155#issuecomment-162403607)

## 2.0.2 - 2015-12-06

### Fixed

- [Button, ComboBox, DatePicker, Rating, Select, TextInput, Toggle] Fix styling issues when the component has a complex React id generated through API's like `createFragment` [c73d06d](https://github.com/nikgraf/belle/commit/c73d06d9ec6f47661657b42e86a6c3a75cc0dc70)

## 2.0.0 - 2015-11-28

### Changed
- Require React 0.14.0 [#189](https://github.com/nikgraf/belle/issues/189)
- [Button, Card, Choice, ComboBox, Option, Placeholder, Rating, Select, Separator, TextInput, Toggle] Removed the 'Belle ' prefix from displayName
- [Button] Prevent sticky hover for touch devices by leveraging mouseEnter & mouseLeave instead of the :hover pseudo class [8f590e2](https://github.com/nikgraf/belle/commit/8f590e27bf0da53828ff7d8fdac32d759e480aa5)
- [Toggle] Hide Webkit tap highlight on touch [5faf2c5](https://github.com/nikgraf/belle/commit/5faf2c56a05f461b20795fc4c097d5259aca3e54)

### Added
- [DatePicker] Add DatePicker component

```
<DatePicker />
```

- [Spinner] Add Spinner component

```
<Spinner />
```

### Fixed
- [Rating] Fix styling the wrapper [82087cb](https://github.com/nikgraf/belle/commit/82087cb220253486e31269c8a989e9cf26fec18e)
- [Select] Fix rendering the selected option with a falsy value e.g. 0 [f0e5611](https://github.com/nikgraf/belle/commit/f0e5611cfce2da3833ed50bd7ac542501ae2999a)
- [Select] Fix hovering and selecting Options with values of type Number [09c52e3](https://github.com/nikgraf/belle/commit/09c52e3ad9f515218fb6fcc922ffcb55c52809a8)

```
<Select>
  <Option value={ 0 }>Zero</Option>
  <Option value={ 1 }>One</Option>
</Select>
```

- [ComboBox] Fix hovering and selecting Options with values of type Number [64a2797](https://github.com/nikgraf/belle/commit/64a27974ecfa6f327063810484e1150d94276699)
- [TextInput] fix calculating the height in case a defaultValue was provided and props were updated [f1a9bc2](https://github.com/nikgraf/belle/commit/f1a9bc2ee0c3fdb74daf805b854fa0c20a249439)

### Removed
- [Rating] Removed the `resetValue` function. Reseting should accomplished by making the Rating a controlled component and passing in the value as `undefined`. See an example below:

```
export default class extends Component {

  state = {
    ratingValue: 3
  }

  reset() {
    this.setState({ ratingValue: undefined });
  }

  render() {
    return (
      <div>
        <button onClick={ ::this.reset() }></button>
        <Rating value={ this.state.ratingValue } />
      </div>
    );
  }
}
```

## 1.2.2 - 2015-09-04

### Fixed
- Support for React 0.14.0-beta3 [#178](https://github.com/nikgraf/belle/pull/178)
- [Toggle] Fix missing base-styles for Toggle [#179](https://github.com/nikgraf/belle/issues/179)

## 1.2.1 - 2015-08-26

### Fixed
- [Combobox, Select] Improve children manipulation to make Belle work with Radium [#170](https://github.com/nikgraf/belle/issues/170)

## 1.2.0 - 2015-08-19

### Added
- [TextInput] Added minRows and maxRows properties [#146](https://github.com/nikgraf/belle/pull/146)

```
<TextInput minRows={ 5 } placeholder="Please fill in description …">
```

### Fixed
- [Select] Avoid printing a PropTypes warning when a Select has less than 2 children. [#147](https://github.com/nikgraf/belle/issues/147)
- [Select] Allowing to a mix single components together with arrays of components as children. [#161](https://github.com/nikgraf/belle/issues/161)

```
<Select>

components/Base.js
import React from 'react';
import PropTypes from 'prop-types';
import { Card } from 'belle';
import { Link } from 'react-router';
import Column from './Column';

export default React.createClass({

  propTypes: {
    route: PropTypes.any.isRequired,
    children: PropTypes.any.isRequired,
    location: PropTypes.object,
  },

  childContextTypes: {
    viewport: PropTypes.any,
  },

  getInitialState() {
    return {
      viewport: this._getRetrieveViewport(),
    };
  },

  getChildContext() {
    return {
      viewport: this._getRetrieveViewport(),
    };
  },

  componentWillMount() {
    window.addEventListener('resize', this._triggerResizeMixinCallback);
  },

  componentWillUnmount() {
    window.removeEventListener('resize', this._triggerResizeMixinCallback);
  },

  _getRetrieveViewport() {
    return {
      width: document.documentElement.clientWidth,
      height: document.documentElement.clientHeight,
    };
  },

  _triggerResizeMixinCallback() {
    this.setState({
      viewport: this._getRetrieveViewport(),
    });
  },

  _renderHomeHeader() {
    return (
      <header style={{ background: 'linear-gradient(145deg, #0A202D, #258AB2)', width: '100%', marginBottom: 40 }}>
        <Column
          smallScreenStyle={{ width: '100%', padding: '0 20px' }}
          mediumScreenStyle={{ margin: '0 auto', width: 910, paddingLeft: 160, paddingRight: 110 }}
        >
          <Link to="/">
            <h1 style={{ fontSize: 80,
                         margin: 0,
                         paddingTop: 80,
                         color: '#FFF',
                         fontFamily: '"Trebuchet MS", Helvetica, sans-serif',
                         position: 'relative',
                       }}
            >
              <div style={{ borderRadius: 100,
                            width: 100,
                            height: 100,
                            border: '4px solid #fff',
                            paddingLeft: 18,
                            display: 'block',
                            float: 'left',
                            marginRight: 12, }}
              >
                <img
                  src="images/belle_logo.png"
                  style={{ width: 62, height: 88 }}
                />
              </div>
              <span style={{ marginTop: -6,
                             display: 'block',
                             float: 'left', }}
              >
                Belle
              </span>
            </h1>
          </Link>
          <p style={{ fontSize: 22, marginTop: 0, paddingTop: 20, color: '#FEFEFE', clear: 'both' }}>
            Configurable React Components with great UX
          </p>
          <iframe
            src="https://ghbtns.com/github-btn.html?user=nikgraf&repo=belle&type=star&count=true&size=large"
            frameBorder="0"
            scrolling="0"
            width="180px"
            height="30px"
            style={{ marginBottom: 20, }}
          />
        </Column>
      </header>
    );
  },

  _renderDefaultHeader() {
    return (
      <header style={{ background: 'linear-gradient(150deg, #0A202D, rgb(1, 95, 130))', width: '100%', marginBottom: 40 }}>
        <Column
          smallScreenStyle={{ width: '100%', padding: '0 20px' }}
          mediumScreenStyle={{ margin: '0 auto', width: 910, paddingRight: 110 }}
        >
          <iframe
            src="https://ghbtns.com/github-btn.html?user=nikgraf&repo=belle&type=star&count=true&size=medium"
            frameBorder="0"
            scrolling="0"
            width="100px"
            height="30px"
            style={{ float: 'right', marginTop: 20 }}
          />

          <Link style={{ display: 'inline' }} to="/">
            <h1 style={{ fontSize: 24, margin: 0, padding: '10px 0', color: '#FFF', fontFamily: '"Trebuchet MS", Helvetica, sans-serif' }}>
              <div style={{ borderRadius: 26,
                            width: 26,
                            height: 26,
                            border: '2px solid #fff',
                            paddingLeft: 5,
                            display: 'block',
                            float: 'left',
                            marginRight: 6,
                            marginTop: 6, }}
              >
                <img
                  src="images/belle_logo.png"
                  style={{ height: 19, top: -7, position: 'relative' }}
                />
              </div>
              Belle
            </h1>
          </Link>
        </Column>
      </header>
    );
  },

  render() {
    const cardContentStyle = (this.state.viewport.width <= 480) ? { padding: 20 } : {};

    return (<div style={{ background: '#F7F7F7' }}>

      { this.props.location.pathname === '/' ? this._renderHomeHeader() : this._renderDefaultHeader() }

      <Column
        smallScreenStyle={{ width: '100%' }}
        mediumScreenStyle={{ margin: '0 auto', width: 910, paddingRight: 110 }}
      >

        <Column
          smallScreenStyle={{ width: '100%', padding: '0 40px' }}
          mediumScreenStyle={{ float: 'left', width: 160 }}
          className="navigation"
        >
          <ul style={{ listStyleType: 'none', paddingLeft: 0, marginTop: 0 }}>
            <li>
              <Link style={{ display: 'block', padding: '3px 0' }} to="/getting-started">Getting Started</Link>
            </li>
          </ul>
          <span style={{ color: '#738088' }} >Components</span>
          <ul style={{ listStyleType: 'none', paddingLeft: 0 }}>
            <li>
              <Link style={{ display: 'block', padding: '3px 0' }} to="/component/button">Button</Link>
            </li>
            <li>
              <Link style={{ display: 'block', padding: '3px 0' }} to="/component/card">Card</Link>
            </li>
            <li>
              <Link style={{ display: 'block', padding: '3px 0' }} to="/component/combo-box">ComboBox</Link>
            </li>
            <li>
              <Link style={{ display: 'block', padding: '3px 0' }} to="/component/date-picker">DatePicker</Link>
            </li>
            <li>
              <Link style={{ display: 'block', padding: '3px 0' }} to="/component/select">Select</Link>
            </li>
            <li style={{ paddingLeft: 10 }}>
              <Link style={{ display: 'block', padding: '1px 0' }} to="/component/option">Option</Link>
            </li>
            <li style={{ paddingLeft: 10 }}>
              <Link style={{ display: 'block', padding: '1px 0' }} to="/component/placeholder">Placeholder</Link>
            </li>
            <li style={{ paddingLeft: 10 }}>
              <Link style={{ display: 'block', padding: '1px 0' }} to="/component/separator">Separator</Link>
            </li>
            <li>
              <Link style={{ display: 'block', padding: '4px 0' }} to="/component/rating">Rating</Link>
            </li>
            <li>
              <Link style={{ display: 'block', padding: '4px 0' }} to="/component/spinner">Spinner</Link>
            </li>
            <li>
              <Link style={{ display: 'block', padding: '4px 0' }} to="/component/text-input">TextInput</Link>
            </li>
            <li>
              <Link style={{ display: 'block', padding: '4px 0' }} to="/component/toggle">Toggle</Link>
            </li>
            <li style={{ paddingLeft: 10 }}>
              <Link style={{ display: 'block', padding: '1px 0' }} to="/component/choice">Choice</Link>
            </li>
          </ul>
          <span style={{ color: '#738088' }} >Guides</span>
          <ul style={{ listStyleType: 'none', paddingLeft: 0 }}>
            <li>
              <Link style={{ display: 'block', padding: '3px 0' }} to="/guide/introducing-belle">Introducing Belle</Link>
            </li>
            <li>
              <Link style={{ display: 'block', padding: '3px 0' }} to="/guide/form-components">Form Components</Link>
            </li>
          </ul>
          <span style={{ color: '#738088' }} >General</span>
          <ul style={{ listStyleType: 'none', paddingLeft: 0 }}>
            <li>
              <Link style={{ display: 'block', padding: '3px 0' }} to="/configuration">Configuration / Styling</Link>
            </li>
            <li>
              <Link style={{ display: 'block', padding: '3px 0' }} to="/philosophy">Philosophy</Link>
            </li>

            <li>
              <Link style={{ display: 'block', padding: '3px 0' }} to="/about">About</Link>
            </li>
            <li>
              <a style={{ display: 'block', padding: '3px 0' }} href="https://github.com/nikgraf/belle" target="_blank">Code on Github</a>
            </li>
            <li>
              <a style={{ display: 'block', padding: '3px 0' }} href="https://github.com/nikgraf/belle/issues" target="_blank">Report an Issue</a>
            </li>
          </ul>
        </Column>

        <Column
          smallScreenStyle={{ width: '96%', margin: '0 auto' }}
          mediumScreenStyle={{ float: 'left', width: 640 }}
        >
          <Card style={ cardContentStyle }>
            { this.props.children }
          </Card>
        </Column>
        <div style={{ clear: 'left' }}></div>
      </Column>
      <footer style={{ clear: 'both', textAlign: 'center', paddingTop: 40, paddingBottom: 60 }}>
        <Column
          smallScreenStyle={{}}
          mediumScreenStyle={{ paddingLeft: 160 }}
        >
          Built with
          <span style={{ color: 'rgb(200, 0, 0)', fontSize: 22, position: 'relative', top: 3 }}>
            &nbsp;&#x2764;&nbsp;
          </span>
          on Planet Earth :)
        </Column>
      </footer>
    </div>);
  },
});
